go-controller/pkg/ovn/egressip.go

```diff
@@ -2099,11 +2101,34 @@ func (oc *Controller) checkEgressNodesReachability() {
 }
 
 func (oc *Controller) isReachable(node *egressNode) bool {
-	for _, ip := range node.mgmtIPs {
-		if dialer.dial(ip) {
-			return true
-		}
-	}
+	var maxRetryCnt int
+	var retryTimeOut, initialRetryTimeOut time.Duration
+
+	switch oc.eIPC.egressIPNodeReachConfig.EgressIPReachabiltyTimeout {
+	// Check if we need to do node reachability check
+	case 0:
+		return true
+	case 1:
+		maxRetryCnt = 2
+		initialRetryTimeOut = time.Second
+		retryTimeOut = 100 * time.Millisecond
+	default:
```

Comment:
its approximation in attempt to avoid b2b attempts and in the sametime didn't want to reduce initial timeout from what its set to currently

Reply:
- 1sec timeout setting
```
1 intf, it took: 1000ms
2 intf, it took: 1000ms
3 intf, it took: 998ms
4 intf, it took: 1000ms
```
- 5sec timeout setting
```
1 intf, it took: 5000ms
2 intf, it took: 5000ms
3 intf, it took: 5000ms
4 intf, it took: 5000ms
```

Reply:
So, just to reiterate my concerns and to argue one last time for running the dials in the background at fixed intervals. Either with a ticker [0] or without [1], and then I will remain silent.

The dialer also returns `false` if we see a syscall.EHOSTUNREACH error:
> https://github.com/ovn-org/ovn-kubernetes/blob/fdd81293dc0008971fa764798d85fcf4028270e6/go-controller/pkg/ovn/egressip.go#L2158

An EHOSTUNREACH error can be returned immediately, so the `min(execution_time(dialer.dial())) ~ 0 ms~. 

You are always only testing code paths:
https://github.com/ovn-org/ovn-kubernetes/blob/fdd81293dc0008971fa764798d85fcf4028270e6/go-controller/pkg/ovn/egressip.go#L2156
https://github.com/ovn-org/ovn-kubernetes/blob/fdd81293dc0008971fa764798d85fcf4028270e6/go-controller/pkg/ovn/egressip.go#L2162

But you are never testing the return in line:
https://github.com/ovn-org/ovn-kubernetes/blob/fdd81293dc0008971fa764798d85fcf4028270e6/go-controller/pkg/ovn/egressip.go#L2159

Which is what you can test if you run the method isolated on your local system, use iptables:
~~~
sudo iptables -I OUTPUT --protocol tcp --dport 9 -j DROP
~~~

Then you get the min runtime; which will be `~ sum(sleep())`.

How can an unreachable error happen? We dial into port 9:
~~~
conn, err := net.DialTimeout("tcp", net.JoinHostPort(ip.String(), "9"), timeout)
~~~

It's also documented in the original function's comments:
https://github.com/openshift/sdn/blob/master/pkg/network/common/egressip.go#L710
~~~
if the node is offline, the
// attempt will either time out with no response, or else return "no route to host" (and
// we will return false).
~~~

The reachability test is done against the nodess' node-subnet IPs: `k8s.ovn.org/node-subnets: '{"default":"10.244.1.0/24"}'`
https://github.com/ovn-org/ovn-kubernetes/blob/1af0a7bee7ab2b9295450736866c91c001d4965a/go-controller/pkg/ovn/egressip.go#L1704

[0] https://go.dev/play/p/qFx-_qEsU1T?v=goprev
The ticker comes with its own drawbacks, so maybe you could replace it with a for loop that runs the dials in the background, sleeps, and another.
I'm not sure if this will leak go routines.

[1] https://go.dev/play/p/NR10srt7MWt?v=goprev
This example does it without tickers and it also shouldn't leak any go routines.

Reply:
the config knob is ONLY to configure timeout duration for the dial() as the name indicates `egressip-reachability-timeout`, so you are right if dial() get `hostunreachable` to the svc on port 9 that knob will be ignored. and I don't think its also correct trying to treat this as timeout not as retrying here with hostunreachable won't ever converge so better to fail fast IMO but I will defer to @trozet or @alexanderConstantinescu

Reply:
Maybe I am asking a stupid question and missing some context as to what we're trying to solve here, but: why can't we just do _exactly_ what openshift-sdn does? :)

Reply:
the PR is doing what osdn does but instead of hardcoding the timeout we made a config knob which can allow us to disable the check completely

Reply:
we cant just do what openshift-sdn does because we need to be more flexible as some use cases call for disabling, some want short timeouts, some want longer. The config option that Mohamed used is an overall timeout to consider a node as unreachable. How we determine that reachability under the covers is up to us. 

I'm not sure what you are getting at @andreaskaris with regards to ticker...I dont really understand your post. port 9 is a special port: https://www.speedguide.net/port.php?port=9

Reply:
Dunno how else to make that clear other than annotating minimum runtimes per line:
~~~
	for retryCnt := 0; retryCnt < maxRetryCnt; retryCnt++ {    0
		for _, ip := range node.mgmtIPs {     0
			if dialer.dial(ip, timeout) {         0     (!!! )
				return true      ---> never reached
			}
		}
		time.Sleep(100 * time.Millisecond)    100 ms
		timeout = retryTimeOut     0
	}
~~~

Reply:
If you want to declare a node down after 10 seconds, there are 2 solutions:
* take the start time, run dialer.dial in a loop, and every iteration check if endtime + 10 seconds > start time; that gives a rough 10 second node down time, but you don't know how often you run the dialer.dial and it won't be on fixed intervals (every run of dialer.dial can take from 0 to timeout seconds)
~~~
	timeout := initialRetryTimeOut
	endTime := time.Now().Add(time.Second * time.Duration(oc.eIPC.egressIPNodeReachConfig.EgressIPReachabiltyTimeout))
	for time.Now().Before(endTime) {
		for _, ip := range node.mgmtIPs {
			if dialer.dial(ip, timeout) {
				return true
			}
		}
		time.Sleep(100 * time.Millisecond)
	}
~~~
* run dialer.dial in the background at fixed intervals  (by "ticking" for example every second)
https://go.dev/play/p/4OriX5YkNjH?v=goprev